File: sources/LiquidSense Release/me/aquavit/liquidsense/module/modules/ghost/BowAimbot.java
package me.aquavit.liquidsense.module.modules.ghost;

import me.aquavit.liquidsense.utils.entity.EntityUtils;
import me.aquavit.liquidsense.event.EventTarget;
import me.aquavit.liquidsense.event.events.Render3DEvent;
import me.aquavit.liquidsense.event.events.UpdateEvent;
import me.aquavit.liquidsense.module.Module;
import me.aquavit.liquidsense.module.ModuleCategory;
import me.aquavit.liquidsense.module.ModuleInfo;
import me.aquavit.liquidsense.utils.client.RotationUtils;
import me.aquavit.liquidsense.utils.render.RenderUtils;
import me.aquavit.liquidsense.value.BoolValue;
import me.aquavit.liquidsense.value.FloatValue;
import me.aquavit.liquidsense.value.ListValue;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.item.ItemBow;
import net.minecraft.item.ItemStack;

import java.awt.*;
import java.util.Comparator;
import java.util.List;
import java.util.function.ToDoubleFunction;
import java.util.stream.Collectors;

@ModuleInfo(name = "BowAimbot", description = "Automatically aims at players when using a bow.", category = ModuleCategory.GHOST)
public class BowAimbot extends Module {

    private final BoolValue silentValue = new BoolValue("Silent", true);
    private final BoolValue predictValue = new BoolValue("Predict", true);
    private final BoolValue throughWallsValue = new BoolValue("ThroughWalls", false);
    private final FloatValue predictSizeValue = new FloatValue("PredictSize", 2.0f, 0.1f, 5.0f);
    private final ListValue priorityValue = new ListValue("Priority", new String[]{"Health", "Distance", "Direction"}, "Direction");
    private final BoolValue markValue = new BoolValue("Mark", true);
    private Entity target;

    @Override
    public void onDisable() {
        this.target = null;
    }

    @EventTarget
    public void onUpdate(UpdateEvent event) {
        target = null;
        ItemStack itemStack = mc.thePlayer.getItemInUse();
        if ((itemStack != null ? itemStack.getItem() : null) instanceof ItemBow) {
            Entity entity = getTarget(throughWallsValue.get(), priorityValue.get());
            if (entity == null)return;
            target = entity;
            RotationUtils.faceBow(target, silentValue.get(), predictValue.get(), predictSizeValue.get());
        }
    }

    @EventTarget
    public void onRender3D(Render3DEvent event) {
        if (target != null && markValue.get())
            RenderUtils.drawPlatform(target, new Color(37, 126, 255, 70));
    }

    private Entity getTarget(boolean throughWalls, String priorityMode) {
        List<Entity> targets = mc.theWorld.loadedEntityList.stream().filter(it -> it instanceof EntityLivingBase && EntityUtils.isSelected(it, true,false) &&
                (throughWalls || mc.thePlayer.canEntityBeSeen(it))).collect(Collectors.toList());;

        switch (priorityMode.toLowerCase()){
            case "distance":
                targets.sort(Comparator.comparingDouble(value -> mc.thePlayer.getDistanceToEntity(value)));
                break;
            case "direction":
                targets.sort(Comparator.comparingDouble((ToDoubleFunction<? super Entity>) RotationUtils::getRotationDifference));
                break;
            case "health":
                targets.sort(Comparator.comparingDouble(it -> ((EntityLivingBase)it).getHealth()));
                break;
            default:
                targets = null;

        }

        return targets != null ? targets.get(0) : null;
    }

    public boolean hasTarget() {
        return target != null && mc.thePlayer.canEntityBeSeen(target);
    }
}

File: sources/Kevin Reborn/kevin/module/modules/combat/BowAimbot.kt
/*
 * This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package kevin.module.modules.combat

import kevin.event.EventTarget
import kevin.event.Render3DEvent
import kevin.event.UpdateEvent
import kevin.module.*
import kevin.utils.EntityUtils
import kevin.utils.RenderUtils
import kevin.utils.RotationUtils
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityLivingBase
import net.minecraft.item.ItemBow
import java.awt.Color

class BowAimbot : Module("BowAimbot", "Automatically aims at players when using a bow.", category = ModuleCategory.COMBAT) {
    private val silentValue = BooleanValue("Silent", true)
    private val predictValue = BooleanValue("Predict", true)
    private val throughWallsValue = BooleanValue("ThroughWalls", false)
    private val predictSizeValue = FloatValue("PredictSize", 2F, 0.1F, 5F)
    private val priorityValue = ListValue("Priority", arrayOf("Health", "Distance", "Direction"), "Direction")
    private val markValue = BooleanValue("Mark", true)

    private var target: Entity? = null

    override fun onDisable() {
        target = null
    }
    @EventTarget
    fun onUpdate(event: UpdateEvent) {
        target = null
        if (mc.thePlayer.itemInUse?.item is ItemBow) {
            val entity = getTarget(throughWallsValue.get(), priorityValue.get()) ?: return
            target = entity
            RotationUtils.faceBow(target, silentValue.get(), predictValue.get(), predictSizeValue.get())
        }
    }
    @EventTarget
    fun onRender3D(event: Render3DEvent) {
        if (target != null && !priorityValue.get().equals("Multi", ignoreCase = true) && markValue.get())
            RenderUtils.drawPlatform(target, Color(37, 126, 255, 70))
    }
    private fun getTarget(throughWalls: Boolean, priorityMode: String): Entity? {
        val targets = mc.theWorld.loadedEntityList.filter {
            it is EntityLivingBase && EntityUtils.isSelected(it, true) &&
                    (throughWalls || mc.thePlayer.canEntityBeSeen(it))
        }
        return when (priorityMode.toUpperCase()) {
            "DISTANCE" -> targets.minByOrNull { mc.thePlayer.getDistanceToEntity(it) }
            "DIRECTION" -> targets.minByOrNull { RotationUtils.getRotationDifference(it) }
            "HEALTH" -> targets.minByOrNull { (it as EntityLivingBase).health }
            else -> null
        }
    }
    fun hasTarget() = target != null && mc.thePlayer!!.canEntityBeSeen(target!!)
}